### PR TITLE
Added new win_package_msp module.

### LINKED ISSUE #22789: support for msp files in win_package
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
win_package.ps1

##### SUMMARY
Support for MSP files.

Currently the win_package module is missing installation of patches in the form of the .msp files. Currently the module only supports .exe and .msi files to install packages. 
I think a new variable "patch_id" is in order, at least to support uninstall of patches.

The primary commands are:
Install msp: msiexe.exe /p “<msp file name with path>” /qn
uninstall msp: msiexec.exe /I <ProductGUID> MSIPATCHREMOVE=<PatchGUID> /qn

### PR DESCRIPTION
**Pull request by Finzzownt**
##### SUMMARY
Addition of new Windows module win_msp. Features installation or removal of Microsoft Installer Patch files (.msp).

Fixes #22789

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
win_msp

##### ADDITIONAL INFORMATION
Use win_msp to install or remove patches (.msp). This module complements the win_package module, which is used to install or remove applications.

### CODE REVIEW

File: lib/ansible/modules/windows/win_msp.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    - One or more return codes from the package installation that indicates
+      success.
+    - A return code of C(3010) usually means that a reboot is required, the
+      I(reboot_required) return value is set if the return code is C(3010).
+    type: list
+    default: [0, 3010]
+  path:
+    description:
+    - Location of the MSP to be installed, e.g. I(state=present).
+    - This package can either be on the local file system, network share or a
```
Comment (it-praktyk) [inline lib/ansible/modules/windows/win_msp.py]: Implementation of downloading a path file - especially from http(s) - shouldn't be, IMHO, the part of that module. Please use [win_get_url](https://docs.ansible.com/ansible/latest/modules/win_get_url_module.html#win-get-url-module) instead.

File: lib/ansible/modules/windows/win_msp.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+# this is a windows documentation stub. Actual code lives in the .ps1
+# file of the same name.
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: win_msp
```
Comment (it-praktyk) [inline lib/ansible/modules/windows/win_msp.py]: What do you think about renaming the module to the `win_package_msp`?

**Review by it-praktyk — CHANGES_REQUESTED** [review]
Lack of integration tests :-(

File: lib/ansible/modules/windows/win_msp.py (hunk 1 above)
Comment (Finzzownt) [reply lib/ansible/modules/windows/win_msp.py]: I agree totally. For instance I've extended the win_get_url module to support downloading and checksum validation. One would argue why not extend that functionality in win_msp too? This was a refactor of the previous win_msi module. It's active in a production environment for six months now, and only now I found the time to contribute. By now it's outdated I noticed. Okay, long story short: I'll remove http(s) support and refer to win_get_url in the documentation.

File: lib/ansible/modules/windows/win_msp.py (hunk 2 above)
Comment (Finzzownt) [reply lib/ansible/modules/windows/win_msp.py]: Ah, good point, as in the comment below, it was rework on the win_msi module. So win_msp was not a bad name back then (April 2018). I think renaming is better. This thread has a mind about it possibly: https://github.com/ansible/ansible/issues/22789, see if anyone else has objections.

**Comment by Finzzownt** [comment]
> Lack of integration tests :-(

Yeah, just discovered the topic on the net. Stating that the code is flawless doesn't help I guess :-)
I'll dive into it. Thanks.

**Comment by jborean93** [comment]
I think we've been on a consensus that win_package should handle these types of packages going forward. I think that code base may need a slight tweak to add future package types a bit easier but a lot of the code here is duplicated from there. I think we should look into adding this feature to win_package instead of having it as a separate module.

**Comment by jhawkesworth** [comment]
@Finzzownt looking at comments above it might be easier to create a new PR to add .msp support to win_package rather than modify this PR.  Shall we close this one?

**Comment by Finzzownt** [comment]
@jhawkesworth Agreed. I'll start merging win_msp into win_package.

**Comment by jborean93** [comment]
Just going through older PRs and based on the conversation above we think the best path forward is to merge this functionality into win_package instead. I'm am going to close this one for now.

**Pull request closed without merging**